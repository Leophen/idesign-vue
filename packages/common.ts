import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
