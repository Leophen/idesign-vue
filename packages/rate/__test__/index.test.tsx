import { vi } from 'vitest';
import { Rate } from '../index';
import { render } from '../../common'

describe('Rate 组件测试', () => {
  it('create', () => {
    const wrapper = render(<Rate />);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('value', () => {
    const wrapper = render(<Rate modelValue={3} />);
    expect(wrapper.findAll('.i-rate-star')[2].find('.i-icon').attributes('style')).toMatch('color: rgb(245, 219, 77)');
    expect(wrapper.findAll('.i-rate-star')[3].find('.i-icon').attributes('style')).not.toMatch('color: rgb(245, 219, 77)');
  });

  it('defaultValue', () => {
    const wrapper = render(<Rate defaultValue={3} />);
    expect(wrapper.findAll('.i-rate-star')[2].find('.i-icon').attributes('style')).toMatch('color: rgb(245, 219, 77)');
    expect(wrapper.findAll('.i-rate-star')[3].find('.i-icon').attributes('style')).not.toMatch('color: rgb(245, 219, 77)');
  });

  it('onChange', () => {
    const changeFn = vi.fn();
    const wrapper = render(<Rate onChange={changeFn} />);
    wrapper.findAll('.i-rate-star')[0].trigger('mousedown');
    expect(changeFn).toHaveBeenCalled();
  });

  it('readonly', () => {
    const changeFn = vi.fn();
    const wrapper = render(<Rate readonly onChange={changeFn} />);
    expect(wrapper.classes()).toContain('i-rate__readonly');
    wrapper.findAll('.i-rate-star')[0].trigger('mousedown');
    expect(changeFn).toBeCalledTimes(0);
  });

  it('allowHalf', () => {
    const wrapper = render(<Rate allowHalf />);
    expect(wrapper.find('.i-rate-star__first').exists()).toBe(true);
  });

  it('count', () => {
    const wrapper = render(<Rate count={10} />);
    expect(wrapper.findAll('.i-rate-star').length).toBe(10);
  });

  it('activeColor voidColor', () => {
    const wrapper = render(<Rate defaultValue={3} activeColor="blue" voidColor="red" />);
    expect(wrapper.findAll('.i-rate-star')[2].find('.i-icon').attributes('style')).toMatch('color: blue');
    expect(wrapper.findAll('.i-rate-star')[3].find('.i-icon').attributes('style')).toMatch('color: red');
  });

  it('activeIcon voidIcon', () => {
    const wrapper = render(<Rate defaultValue={3} activeIcon="BellFill" voidIcon="Bell" />);
    expect(wrapper.find('.i-icon').classes()).toContain('icon-BellFill');
  });
});

