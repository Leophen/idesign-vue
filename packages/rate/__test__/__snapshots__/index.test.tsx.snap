// Vitest Snapshot v1

exports[`Rate ç»„ä»¶æµ‹è¯• > create 1`] = `
<div
  class="i-rate"
>
  
  <div
    class="i-rate-star"
  >
    <i
      class="i-icon icon-StarFill"
      style="font-size: 20px;"
    />
  </div>
  <div
    class="i-rate-star"
  >
    <i
      class="i-icon icon-StarFill"
      style="font-size: 20px;"
    />
  </div>
  <div
    class="i-rate-star"
  >
    <i
      class="i-icon icon-StarFill"
      style="font-size: 20px;"
    />
  </div>
  <div
    class="i-rate-star"
  >
    <i
      class="i-icon icon-StarFill"
      style="font-size: 20px;"
    />
  </div>
  <div
    class="i-rate-star"
  >
    <i
      class="i-icon icon-StarFill"
      style="font-size: 20px;"
    />
  </div>
  
</div>
`;
